File: lab_sheet_2/q1.cpp
// 19000073
// Lab sheet 2

#include <iostream>
#include <map>
#include <vector>
#include <deque>
using namespace std;

class Graph {
    public:
    map< char, deque<char> > vertices;    
    void insertVertex (char v) {
        deque<char> newEdges;
        vertices[v] = newEdges;
    }
    void insertEdge (char from, char to) {
        if (vertices.count(from) == 0 || vertices.count(to) == 0) {
            cout << "Couldn't insert vertex make sure the given vertices are in the graph" << endl;
            return;
        }
        vertices[from].push_back(to);
    }
    // Question 01 - Checking if theres a path between two vertices
    bool checkPath (char from, char to) {
        if (vertices.count(from) == 0 || vertices.count(to) == 0) {
            cout << "Couldn't check for paths, make sure the given vertices are in the graph" << endl;
            return false;
        }
        deque<char> vertexQ = vertices[from];
        vector<char> visited;
        while (!vertexQ.empty()) {
            char current = vertexQ.front();
            vertexQ.pop_front();
            visited.push_back(current);
            if (current == to) {
                return true;
            }
            for (int i = 0; i < vertices[current].size(); i++) {
                bool isVisited = false;
                for (int j = 0; j < visited.size(); j++) {
                    if (vertices[current][i] == visited[j]) {
                        isVisited = true;
                        break;
                    }
                }
                if (!isVisited) {
                    vertexQ.push_back(vertices[current][i]);
                }
            }
        }
        return false;
    }
    // Question 02 - Checking if theres a cycle in the graph
    bool checkCycle () {
        for (map< char, deque<char> >::iterator i = vertices.begin(); i != vertices.end(); i++) {
            if (checkPath(i->first, i->first)) return true;
        }
        return false;
    }
    // Question 03 - Counting number of paths between two vertices
    int noOfPaths (char from, char to) {
        if (vertices.count(from) == 0 || vertices.count(to) == 0) {
            cout << "Couldn't check for paths, make sure the given vertices are in the graph" << endl;
            return -1;
        }
        if (this->checkCycle()) {
            cout << "The graph contains a cycle, couldn't count the no of paths" << endl;
            return -1;
        }
        int pathCount = 0;
        deque<char> vertexQ = vertices[from];
        while (!vertexQ.empty()) {
            char current = vertexQ.front();
            vertexQ.pop_front();
            if (current == to) {
                pathCount++;
                continue;
            }
            for (int i = 0; i < vertices[current].size(); i++) {
                vertexQ.push_back(vertices[current][i]);
            }
        }
        return pathCount;
    }
    void displayGraph () {
        for (map<char, deque<char> >::iterator i = vertices.begin(); i != vertices.end(); i++) {
            cout << "(" << i->first << ") -> { ";
            deque<char> edges = i->second;
            for (int j = 0; j < edges.size(); j++) {
                cout << edges[j] << " ";
            }
            cout << "}" << endl;
        }
    }
};

int main () {
    Graph g;
    g.insertVertex('a');
    g.insertVertex('b');
    g.insertVertex('c');
    g.insertVertex('d');
    g.insertVertex('e');

    g.insertEdge('a', 'b');
    g.insertEdge('b', 'c');
    g.insertEdge('c', 'a');
    g.insertEdge('d', 'c');
    g.insertEdge('e', 'd');

    g.displayGraph();
    g.checkPath('e', 'a') ? cout << "Path found from e to a!" << endl : cout << "Path not found" << endl;
    g.checkCycle() ? cout << "Cycle found!" << endl : cout << "Cycle not found" << endl;
    g.noOfPaths('e', 'a');

    Graph g1;
    g1.insertVertex('a');
    g1.insertVertex('b');
    g1.insertVertex('c');
    g1.insertVertex('d');
    g1.insertVertex('e');
    g1.insertVertex('f');
    g1.insertVertex('g');

    g1.insertEdge('a', 'b');
    g1.insertEdge('a', 'c');
    g1.insertEdge('b', 'e');
    g1.insertEdge('c', 'd');
    g1.insertEdge('d', 'e');
    g1.insertEdge('c', 'f');
    g1.insertEdge('f', 'e');
    g1.insertEdge('e', 'g');

    g1.displayGraph();
    g1.checkPath('a', 'g') ? cout << "Path found from a to g!" << endl : cout << "Path not found" << endl;
    g1.checkCycle() ? cout << "Cycle found!" << endl : cout << "Cycle not found" << endl;
    cout << "No of paths from a to g: " << g1.noOfPaths('a', 'g') << endl;

    Graph g2;
    g2.insertVertex('a');
    g2.insertVertex('b');
    g2.insertVertex('c');
    g2.insertVertex('d');
    g2.insertVertex('e');
    g2.insertVertex('f');
    g2.insertVertex('g');

    g2.insertEdge('a', 'b');
    g2.insertEdge('a', 'c');
    g2.insertEdge('b', 'e');
    g2.insertEdge('c', 'd');
    g2.insertEdge('d', 'e');
    g2.insertEdge('f', 'e');
    g2.insertEdge('e', 'g');

    g2.displayGraph();
    g2.checkPath('a', 'g') ? cout << "Path found from a to g!" << endl : cout << "Path not found" << endl;
    g2.checkCycle() ? cout << "Cycle found!" << endl : cout << "Cycle not found" << endl;
    cout << "No of paths from a to g: " << g2.noOfPaths('a', 'g') << endl; 

    return 0;
}

File: lab_sheet_3/Q1.py
# 19000073, Lab Sheet 3 - Question 1
from collections import deque


class Graph:
    def __init__(self):
        self.vertices = {}

    # Add vertex Question 1 part 1
    def addVertex(self, v):
        self.vertices[v] = list()

    # Add edge Question 1 part 1
    def addEdge(self, v1, v2):
        self.vertices[v1].append(v2)

    # Remove vertex Question 1 part 2
    def removeVertex(self, v):
        self.vertices.pop(v)
        for u in self.vertices:
            self.vertices[u].remove(v)

    # Find path and length of the path between two vertices
    # Question 1 part 3, 4
    def findPath(self, s, d):
        q = deque()
        q.append(s)
        visited = [s]
        prev = [None] * len(self.vertices)
        while q:
            current = q.popleft()
            for n in self.vertices[current]:
                if n not in visited:
                    q.append(n)
                    visited.append(n)
                    prev[n] = current
        current = d
        path = [current]
        while current:
            current = prev[current]
            path.append(current)
        path.reverse()
        if path[0] == s:
            print("Path found: {} to {}".format(s, d))
            print("- Path: ", end=" ")
            for v in path:
                print(v, end=" ")
            print("")
            print("- Length: {}".format(len(path) - 1))
            return
        print("Path not found: {} to {}".format(s, d))

    def printGraph(self):
        print("-------------------------------")
        for v in self.vertices:
            print("{} -> (".format(v), end=" ")
            for u in self.vertices[v]:
                print(u, end=" ")
            print(")")
        print("-------------------------------")


g = Graph()

# Add vertices
g.addVertex(0)
g.addVertex(1)
g.addVertex(2)
g.addVertex(3)
g.addVertex(4)
g.addVertex(5)
g.addVertex(6)
g.addVertex(7)
g.addVertex(8)
g.addVertex(9)
g.addVertex(10)
g.addVertex(11)
g.addVertex(12)

# Add edges
g.addEdge(0, 7)
g.addEdge(0, 9)
g.addEdge(0, 11)
g.addEdge(9, 8)
g.addEdge(9, 10)
g.addEdge(7, 11)
g.addEdge(7, 6)
g.addEdge(7, 3)
g.addEdge(8, 1)
g.addEdge(8, 12)
g.addEdge(10, 1)
g.addEdge(6, 5)
g.addEdge(3, 4)
g.addEdge(3, 2)
g.addEdge(12, 2)

g.printGraph()
g.findPath(0, 2)
g.findPath(9, 5)
g.findPath(0, 11)
g.findPath(0, 12)
